README.md
## adaptive-linechart :chart_with_upwards_trend:

Version 1.2.0

This is a small library to draw charts in the browser. Why is it "adaptive"? Because it automaticaly
adjust chart to available size depend on min/max values on and timespan. Additional you 
can adjust the layout of the chart. The chart uses the external CSS styles .nowrap .linechart, 
.chart-title, .chart-live-data, .chart-legend, .chart-legend-item, .chart-legend-color. 

Maximal 10 lines. Since version 1.1.0 allowed negative and positive values in randge from 
-9007199254740991 till 9007199254740991. The chart uses the global variable isMobil (Boolean). If
false, live data are shown by "mouse-over". By click on item in the legend hide/show line on chart 
and switch colored tag of item.

The chart accepts data as time-series in the format (string) "Unix-Timestamp/Value;Unix-Timestamp/Value;..".
Additional (chart-util) supplied function to convert CSV (with header) to time-series.

The library uses its own simple logger that writes messages to the console. If needed, you can
rewrite this object and redirect messages to another target.

## Example

This example creates a chart with 3 lines in parent element with ID 'charts':
```js
var arTimeSeries = [
    {
        name: 'Open',
        data: '1565301600/138.6100;1565560800/137.0700;1565647200/136.0500;1565733600/136.3600;1565820000/134.3900'
    },
    {
        name: 'Hight',
        data: '1565301600/139.3800;1565560800/137.8600;1565647200/138.8000;1565733600/136.9200;1565820000/134.5800'
    },
    {
        name: ' Low',
        data:  '1565301600/136.4600;1565560800/135.2400;1565647200/135.0000;1565733600/133.6700;1565820000/132.2500'
    },
];

var lchart =  new LineChart(document.getElementById('charts'),null, null, 'Example Chart-Line', null);

for(let i=0; i < arTimeSeries.length; i++) {
        let line = {
            obj_name: 'Stock preis',
            par_name: arTimeSeries[i].name,
            values: arTimeSeries[i].data
        };
        lchart.addLine(line);
    }
            
lchart.draw();
```

[Live example](http://www.softdorado.com/adaptive-linechart/)

## Public methods

**`LineChart(container, chartId, width, title, className)`** - create an object of type LineChart 
and create HTML code for the chart.
- **container** - Reference to parent element for chart. If null: use BODY as parent element.
- **chartId** - String. Identifier for chart. Optional. If not defined use random integer.
- **width** - Integer. Width in pixel. Optional. Default use available width in parent element.
- **title** - String. Title of chart. Optional. If null: don't create title.
- **className** - String. Class for div of chart. Optional. Default "linechart".

**`addLine(lineData)`** - add data for a new line and automatically create a new item in the legend. 
Return true if no problem detected, otherwise false.

**`lineData`** - The object represents data for one line in the chart.
- **obj_name** - String. Name of object related to the chart. Used in legend.
- **par_name** - String. Name of the parameter that represents the current line. Used in legend.
- **par_id** - Integer. Optional ID of this line. May be used to request data from backend.
- **values** - String. Time-series in format "Unix-Timestamp/Value;Unix-Timestamp/Value;.."

**`draw()`** - draw graphic. Skip lines if properties hidden=true or isEmpty=true.

**`clear()`** - clear graphic. Legend stays visible.

**`resize(difWidth, difHeight)`** - set a new size. E.g. if the size of parent element is changed.
- **difWidth** - Signed integer. Difference in width.
- **difHeight** - Signed integer. Difference in height.

**`isEmpty()`** Return true if all lines are empty

## Utility

**`scvToTimeSeries(csv)`** - process CSV and return array of objects 

`{name: "Name of serie", data: "Unix-Timestamp/Value;Unix-Timestamp/Value"}`. 

Assume the first column contains data-time as string. The CSV may contain many columns. Name of 
serie taken from header. Values may be empty. This function needs the package [Moment.js](https://momentjs.com)

## Contact

If you have any questions, please contact me `adaptive-linechart@my.mail.de`

## License

[MIT](https://opensource.org/licenses/MIT) © Andrej Koslov



CHANGELOG.md
# Changelog

## 1.2.0 (01 Nov 2020)
* New feature:
	- By click on legend item hide/show line on chart.
* Fix bug:
	- In some case label on axe-y below axe-x.

## 1.1.0 (21 May 2020)
Allowed negative and positive values in randge from -9007199254740991 till 9007199254740991.

## 1.0.1 (4 May 2020)
- Add README.md
- Add directory **src** (files for development) and **dist** (ready to use minified files)

## 1.0.0 (26 April 2020)
- First Version



src/chart-util.js
// 
// Collection of small utilities
//

// Create simple logger. Used to log pisible error by processing data of chart.
if(typeof logger != "notdefined") {
    var logger = {
        info: function (message) { console.info(message);},
        warn: function (message) {console.warn(message);},
        error: function (message) {console.error(message);},
        debug: function (message) {console.debug(bessage);}
    };
}

// Add new element with predefined attributes and return reference to new element.
// @param parent - Reference to parent element. If null create new element in <body>.
// @param tag String. Tag of new element. Default <div>
// @param attrSet Object. Contains sub-elements. Each is one attribute of DOM element.
//   id - String. ID of new element. If not defined or is not unique create random. 
//   innerText - String. If defined add as innerText.
//   innerHTML - String- If defined add as innerHTML.
//   Example {id:'chart_legend_item', style: 'display:none; padding:10px;'}
// @return Reference to DOM element.
function addNewElement(parent, tag, attrSet) {
    if (attrSet == null) attrSet = {};

    // Check if id exist and is unique. If not set quasi-random id.
    if (attrSet.id != null) {
        if (document.getElementById(attrSet.id) != null) {
            newElementCounter++;
            attrSet.id = newElementCounter.toString()+ '_'+ new Date().getTime().toString();
        }
    } else {
        newElementCounter++;
        attrSet.id = newElementCounter.toString()+ '_'+ new Date().getTime().toString();
    }

    // Create new element. If type not defined create div.
    if (tag == null) {
        var element = document.createElement('div');
    } else {
        var element = document.createElement(tag);
    }

    // Set attributes
    for(var attribute in attrSet){
        element.setAttribute(attribute, attrSet[attribute]);
    }

    // Add intern text or html code if defined.
    if(attrSet.innerText !=  null) element.innerText = attrSet.innerText;
    if(attrSet.innerHTML !=  null) element.innerHTML = attrSet.innerHTML;

    // Append new element to body or to parent element if it defined and exist
    if (parent == null) {
        document.body.appendChild(element);
    } else {
        parent.appendChild(element);
    }

    return document.getElementById(attrSet.id);
}

// Counter that used in function addNewElement() to generate unique id
var  newElementCounter = 0;

// Calculate average value from array and return float with defined precision
// if at least one item in array is a number else return NaN.
// @param arData - Array. Array of numbers.
// $param precision - Integer. Default 2 digits.
// @return Float or NaN.
function getAvgValue (arData, precision) {

    if(arData == null || arData.length == 0) return NaN;

    if(precision == null || isNaN(precision)) precision = 2;

    var sum = 0;
    var count = 0;

    for(var i=0; i< arData.length; i++) {
        if (!isNaN(arData[i])) {
            sum += arData[i];
            count += 1;
        }
    }

    if(count == 0) {
        return NaN;
    } else {
        return (sum / count).toFixed(precision);
    }
}

// Get random integer with 16 digits.
function getRandomInt () {
    return Math.random() * 100000000000000000;
};

// If variable is string return true, else false.
function isString(variable) {
    return typeof variable === 'string' || variable instanceof String;
}

// Function return object that contains pieces of data-time as strings. Year long, month 01..12,
// day of month 01..31, hour 01..23, min 01..59, sec 01..59, msec 001...999
Date.prototype.separate = function() {

    var dt = {};

    dt.year = this.getFullYear();
    dt.month = this.getMonth() + 1;
    dt.mday = this.getDate();
    dt.hour = this.getHours();
    dt.min = this.getMinutes();
    dt.sec = this.getSeconds();
    dt.msec = this.getMilliseconds();

    if(dt.month < 10)  dt.month = '0' + dt.month;
    if(dt.mday < 10) dt.mday = '0'+ dt.mday;
    if(dt.hour < 10) dt.hour = '0'+ dt.hour;
    if(dt.min < 10) dt.min = '0'+ dt.min;
    if(dt.sec < 10) dt.sec = '0'+ dt.sec;
    if(dt.msec < 10) {
        dt.msec = '00'+ dt.msec;
    } else if (dt.msec < 100) {
        dt.msec = '0'+ dt.msec;
    }

    return dt;
};

// Convert CSV with header to array of time-series. Assume first value is data-time as string. 
// Name of serie take from header. CSV may contains many series (columns). Values may be empty. 
// Each serie is object {name: "Name of serie", data: "Unix-Timestamp/Value;Unix-Timestam/Value"}
// This function need package Moment.js https://momentjs.com
function scvToTimeSeries(csv) {
     
    let arResult = [];
     
    if(csv == null) {return arResult;}
    
    let arLines = csv.split(/\r?\n/);
     
    if(arLines.length < 2) {return arResult;}
    
    // Save header in separate array. Later skeep firt line with header.
    var arHeader = arLines[0].split(',');
     
    // For each column with data (In header first columnt is date-time) create time-serie object
    // and fill it with data from column.
    for(let x=1; x < arHeader.length; x++) {
        
        let obSerie = {name: arHeader[x], data: ""};
        
        // Process all records in one columnt. Skeep first line with header.
        // Process lines in reverse order, because newest crecords on top of csv.
        // If process in usual order chart works also correct! I don't know why : /
        for(let i=arLines.length - 1; i > 0; i--) {
            let arElements = arLines[i].split(',');

            // Skeep if line contains invalid data. Amount of columns in current line and header 
            // should be the same.
            if(arElements.length != arHeader.length)
                continue;
            
            // Build one item in time-serie. Format "data-time/value;" At the end separator between 
            // items ";". 
            obSerie.data += moment(arElements[0]).format('X')+ '/'+ arElements[x]+ ';';
            
        }

        // Remove last separator
        obSerie.data = obSerie.data.slice(0, -1);
        
        // Add new time-serie to array
        arResult.push(obSerie);
    }
     
    return arResult;
}

//
// Extentions existing types
//

example/example.js
document.addEventListener('DOMContentLoaded', function () {
    createCharts();
});

var isMobil = false;
var request;
var charts = [];

// Example time-series. Semicolon separated pairs "Unix-Timestamp/Value;Unix-Timestamp/Value;.."
// The values may be empty. The last element may be withoud value. It Indicate end of timespane
// of this line.
var arTimeSeries = [
    {
        name: 'Example one',
        data: '1565301600/-11.2;1565560800/-9;1565647200/-7.2;1565733600/-5.89;1565820000/-5.2;1565906400/-2.50;1566165600/-1.21;1566252000/0.20;1566338400/-2;1566424800/-4.5;1566511200/-6.2;1566770400/-8.3;1566856800/-10.0;1566943200/-11;1567029600/-13.8;1567116000/-15.8;1567461600/-12.9;1567548000/-8.9;1567634400/-8.94;1567720800/-10.96'
    },
    {
        name: 'Example Two',
        data: '1565301600/11.2;1565560800/6;1565647200/4;1565733600/2;1565820000/-2;1565906400/-3;1566165600/-4;1566252000/-6;1566338400/-7;1566424800/-8;1566511200/-8;1566770400/-5.5;1566856800/-3.6;1566943200/-0.6;1567029600/1.8;1567116000/3.8;1567461600/4.2;1567548000/5.4;1567634400/6.94;1567720800/7.96'
    },
    {
        name: 'Example Two',
        data: '1565301000/0;1565560200/-0.4;1565646600/-1.1;1565733000/-2;1565819400/-0.5;1565905800/-0.01;1566165000/0.2;1566252000/0.3;1566337800/2.2;1566424200/4.2;1566510600/5.2;1566779800/6.1;1566856200/7.4;1566942600/8;1567029000/8.2;1567115400/8.8;1567461000/9.2;1567547400/11.4;1567633800/12.94;1567720200/11.96'
    }
];

function createCharts () {

    // Create first chart in div id "charts"
    let lchart;
    
    lchart = new LineChart(document.getElementById('charts'),null, null, 'Example Chart-Line', null);
             
    for(let i=0; i < arTimeSeries.length; i++) {
        let line = {
            obj_name: 'Chart',
            par_name: arTimeSeries[i].name,
            values: arTimeSeries[i].data
        };
        lchart.addLine(line);
    }
            
    lchart.draw();

    charts.push(lchart);

    // Example convert CSV to time-series
    request = new XMLHttpRequest();
    request.open("GET","example.csv");
    request.addEventListener('load', function(event) {
        if (request.status == 200) {
           
            // Convert SCV to format suitable for chart.
            let arTimeSeries = scvToTimeSeries(request.responseText);
            
            // Create next chart in div id "charts"
            lchart =  new LineChart(document.getElementById('charts'),null, null, 'Example Chart-Line with data from CSV', null);
             
            for(let i=0; i < arTimeSeries.length - 1; i++) {
                let line = {
                    obj_name: 'Stock preis',
                    par_name: arTimeSeries[i].name,
                    values: arTimeSeries[i].data
                };
                lchart.addLine(line);
            }
            
            lchart.draw();

            charts.push(lchart);
        } else {

           console.warn("Request CSV failed status "+ request.statusText+ " Responce "+ request.responseText);
        }
     });

    request.send(); 
}




                        
                        